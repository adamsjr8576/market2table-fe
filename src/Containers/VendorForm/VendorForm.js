import React, { useState, useEffect, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { useDropzone } from 'react-dropzone'
import request from 'superagent';
import './VendorForm.scss';
import { addVendors, addSelectedVendor } from '../../actions';
import images from '../../images/images';
import VendorProductContainer from '../VendorProductContainer/VendorProductContainer';
import { createVendor, createProduct } from '../../apiCalls';

export const VendorForm = () => {
  const [vendorName, setVendorName] = useState('');
  const [vendorDescription, setVendorDescription] = useState('');
  const [vendorImage, setVendorImage] = useState('');
  const [vendorProducts, setVendorProducts] = useState([]);
  const [hasError, setHasError] = useState(false);
  const CLOUDINARY_UPLOAD_PRESET = 'Farmer_Images';
  const CLOUDINARY_UPLOAD_URL = 'https://api.cloudinary.com/v1_1/dlgdlli2u/image/upload';
  const vendor = useSelector(state => state.selectedVendor);
  const dispatch = useDispatch();


  useEffect(() => {
    if (vendor.name) {
      setVendorName(vendor.name);
      setVendorDescription(vendor.description);
      setVendorImage(vendor.image_link);
    }
  }, []);

  const handleImageUpload = (file) => {
    let upload = request.post(CLOUDINARY_UPLOAD_URL)
                        .field('upload_preset', CLOUDINARY_UPLOAD_PRESET)
                        .field('file', file);
    upload.end((err, response) => {
      if (err) {
        console.error(err);
      }
      if (response.body.secure_url !== '') {
        setVendorImage(response.body.secure_url);
      }
    });
  }

  const onDrop = useCallback(acceptedFiles => {
    if (acceptedFiles.length) {
      handleImageUpload(acceptedFiles[0]);
      setHasError(false);
    }
  }, []);

  const {getRootProps, getInputProps, isDragActive, isDragAccept, isDragReject} = useDropzone({accept: 'image/jpeg, image/png, image/jpg', disabled: vendorImage.length > 0, onDrop});

  let image;
  if (vendorImage.length > 0) {
    image = (
      <section className='uploaded-image-container'>
        <img src={vendorImage} alt='photo of farm' className='farmer-image' />
        <button type='button' className='delete-uploaded-image' onClick={() => setVendorImage('')}>X</button>
      </section>
    )
  } else {
    image = (
      <p>View uploaded image here</p>
    )
  }

  const handleSubmitForm = () => {
    if (vendorName.length > 0 && vendorDescription.length > 0) {
      createVendor(vendorName, vendorDescription, vendorImage)
      .then(vendor => {
        if (vendorProducts.length) {
          vendorProducts.forEach(product => {
            createProduct(product.name, product.description, product.price, vendor.data.addVendor.id)
              .then(data => {
                return (<Redirect to='/vendor/account'/>)
              })
              .catch(error => console.log(error))
          });
        }

      })
      .catch(error => console.log(error));
      setVendorName('');
      setVendorDescription('');
      setVendorImage('');
      setVendorProducts([]);
    } else {
      setHasError(true);
    }
  }

  return (
    <section className='vendor-form-main-section'>
      <header className='vendor-form-header'>
        <Link to='/vendor/account' className='link-back-to-vendors' onClick={() => dispatch(addSelectedVendor({}))}>
          <img src={images.undo} className='undo-image' alt='icon of reverse array' />
          <p className='back-to-vendors-p'>Back To Vendors</p>
        </Link>
        <h2 className='vendor-form-title'>Vendor Form</h2>
        <div className='vendor-form-header-space'></div>
      </header>
      <form className='vendor-form-info' id='vendor-form-info'>
        <div className='vendor-form-input-container'>
          <label for='vendor-name-input' className='vendor-name-label'>* Enter Vendor Name:</label>
          <input value={vendorName} onChange={(event) => {setVendorName(event.target.value); setHasError(false);}}
            type='text' className='vendor-name-input' maxlength='35'
            placeholder='Vendor Name...' id='vendor-name-input' />
        </div>
        <div className='vendor-form-input-container'>
          <label for='vendor-description-textarea' className='vendor-description-label'>* Enter Vendor Description:</label>
          <textarea value={vendorDescription} onChange={(event) => {setVendorDescription(event.target.value); setHasError(false);}}
            form='vendor-form-info' name='vendor-description-textarea'
            className='vendor-description-textarea' placeholder='Vender Description...'
            id='vendor-description-textarea' rows='6' columns='25' maxlength='150'>
          </textarea>
        </div>
        <div className='image-uploader-container'>
          <div {...getRootProps()} className='image-dropper'>
            <input {...getInputProps()} />
            {
              isDragActive ?
                <section className='image-dropper-section' style={{backgroundColor: '#EBEEE6'}}>
                  <p className='image-dropper-text'>Drop the file here</p>
                  <img src={images.addImage} className='plus-image-dropper'/>
                  <p className='image-dropper-text'>.jpg, .jpeg, .png only</p>
                </section> :
                <section className='image-dropper-section'>
                  <p className='image-dropper-text'>Drag & drop, or click to select file</p>
                  <img src={images.addImage} className='plus-image-dropper'/>
                  <p className='image-dropper-text'>.jpg - .jpeg - .png only</p>
                </section>
            }
          </div>
          <div className='farmer-image-container'>
            {image}
          </div>
        </div>
      </form>
      <VendorProductContainer products={vendorProducts} setProducts={setVendorProducts}/>
      <section className='submit-vendor-info-button-section'>
        <button className='submit-vendor-info-button' type='button' onClick={handleSubmitForm}><p className='create-vendor-button-p'>Submit Vendor</p><img className='plus-icon-form' src={images.plus} alt='plus sign icon'/></button>
        <p hidden={!hasError} className='vendor-form-error'>*Please include all required fields*</p>
      </section>
    </section>
  )
}

export default VendorForm;
