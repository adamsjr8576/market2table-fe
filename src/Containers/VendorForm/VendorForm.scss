@import './src/variables';

.vendor-form-main-section {
  width: 90%;
  height: 82vh;
  background-color: white;
  margin: auto;
  border-radius: 10px;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-around;
}

.vendor-form-header {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-around;
}

.link-back-to-vendors {
  width: 12%;
  min-width: 130px;
  margin-left: 5px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  text-decoration: none;
  color: black;
}

.link-back-to-vendors:hover {
  .undo-image {
    transform: scale(1.1);
  }
}

.back-to-vendors-p {
  margin: 0;
}

.undo-image {
  height: 20px;
  width: 20px;
}

.vendor-form-title {
  width: 73%;
  font-size: 2.5em;
  margin: 0;
  text-align: center;
  text-decoration: underline;
}

.vendor-form-header-space {
  width: 12%;
}

.vendor-form-info {
  height: 85%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  width: 40%;
  min-width: 300px;
}

.vendor-form-input-container {
  width: 85%;
  display: flex;
  flex-direction: column;
}

.vendor-name-input {
  width: 75%;
  margin-top: 10px;
  border-radius: 20px;
  border: 3px solid $dark_red_border_color;
  background-color: $light_red_fill_color;
  padding: 10px;
  outline: 0;
  font-size: 1em;
}

.vendor-description-textarea {
  resize: none;
  font-size: 1em;
  margin-top: 10px;
  outline: 0;
  border: 3px solid $dark_red_border_color;
  background-color: $light_red_fill_color;
  border-radius: 10px;
  padding: 10px;
}

.image-uploader-container {
  height: 35%;
  width: 90%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.image-dropper {
  width: 45%;
  height: 75%;
  border: 2px dashed black;
  cursor: pointer;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.image-dropper-text {
  width: 90%;
  height: 50%;
  margin: 5px 0;
  text-align: center;
}

.plus-image-dropper {
  margin-top: 10px;
  height: 50px;
  width: 50px;
}

.image-dropper-section {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.farmer-image-container {
  width: 45%;
}

.farmer-image {
  height: 150px;
  width: 150px;
}

.submit-vendor-info-button {
  width: 10%;
  height: 150px;
  margin: 0 20px;
  border-radius: 25px;
  font-size: 2em;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  outline: 0;
  border: 2px solid #3B3D39;
  background-color: #F0F0EF;
}

.submit-vendor-info-button:hover,
.submit-vendor-info-button:focus {
  background-color: #C4D1B2;
  border: 2px solid #5E7838;
}

.submit-vendor-info-button:active {
  transform: scale(.98);
}

.plus-icon-form {
  height: 40px;
  width: 40px;
  margin-top: 10px;
}

.create-vendor-button-p {
  margin: 0;
}
