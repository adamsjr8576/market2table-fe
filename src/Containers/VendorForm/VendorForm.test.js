import React from 'react';
import { shallow } from 'enzyme';
import { VendorForm } from './VendorForm';

const mockState = {
  selectedVendor: {name: 'Rolling in the dough', description: 'just a little bakery rolling in some dough', image_link: 'rollthatdough.png'}
}

jest.mock("react-redux", () => ({
  useSelector: () => mockState,
  useDispatch: () => jest.fn()
}));

describe('VendorForm', () => {

  let wrapper;
  let mockAddVendors = jest.fn();
  let mockAddSelectedVendor = jest.fn();

  it('should match the snapshot', () => {
    wrapper = shallow(<VendorForm
      addVendors={mockAddVendors}
      addSelectedVendor={mockAddSelectedVendor}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
