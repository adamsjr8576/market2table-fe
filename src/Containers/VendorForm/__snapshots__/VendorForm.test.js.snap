// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VendorForm should match the snapshot 1`] = `
<section
  className="vendor-form-main-section"
>
  <header
    className="vendor-form-header"
  >
    <Link
      className="link-back-to-vendors"
      onClick={[Function]}
      to="/vendor/account"
    >
      <img
        alt="icon of reverse array"
        className="undo-image"
        src="undo.png"
      />
      <p
        className="back-to-vendors-p"
      >
        Back To Vendors
      </p>
    </Link>
    <h2
      className="vendor-form-title"
    >
      Vendor Form
    </h2>
    <div
      className="vendor-form-header-space"
    />
  </header>
  <form
    className="vendor-form-info"
    id="vendor-form-info"
  >
    <div
      className="vendor-form-input-container"
    >
      <label
        className="vendor-name-label"
        for="vendor-name-input"
      >
        * Enter Vendor Name:
      </label>
      <input
        className="vendor-name-input"
        id="vendor-name-input"
        maxlength="35"
        onChange={[Function]}
        placeholder="Vendor Name..."
        type="text"
        value=""
      />
    </div>
    <div
      className="vendor-form-input-container"
    >
      <label
        className="vendor-description-label"
        for="vendor-description-textarea"
      >
        * Enter Vendor Description:
      </label>
      <textarea
        className="vendor-description-textarea"
        columns="25"
        form="vendor-form-info"
        id="vendor-description-textarea"
        maxlength="150"
        name="vendor-description-textarea"
        onChange={[Function]}
        placeholder="Vender Description..."
        rows="6"
        value=""
      />
    </div>
    <div
      className="image-uploader-container"
    >
      <div
        className="image-dropper"
        onBlur={[Function]}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <input
          accept="image/jpeg, image/png, image/jpg"
          autoComplete="off"
          multiple={true}
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          tabIndex={-1}
          type="file"
        />
        <section
          className="image-dropper-section"
        >
          <p
            className="image-dropper-text"
          >
            Drag & drop, or click to select file
          </p>
          <img
            className="plus-image-dropper"
            src="addImage.png"
          />
          <p
            className="image-dropper-text"
          >
            .jpg - .jpeg - .png only
          </p>
        </section>
      </div>
      <div
        className="farmer-image-container"
      >
        <p>
          View uploaded image here
        </p>
      </div>
    </div>
  </form>
  <VendorProductContainer
    products={Array []}
    setProducts={[Function]}
  />
  <section
    className="submit-vendor-info-button-section"
  >
    <button
      className="submit-vendor-info-button"
      onClick={[Function]}
      type="button"
    >
      <p
        className="create-vendor-button-p"
      >
        Submit Vendor
      </p>
      <img
        alt="plus sign icon"
        className="plus-icon-form"
        src="plus.png"
      />
    </button>
    <p
      className="vendor-form-error"
      hidden={true}
    >
      *Please include all required fields*
    </p>
  </section>
</section>
`;
